pkg/apis/pipeline/v1beta1/when_validation.go

```diff
@@ -79,12 +78,15 @@ func (wes WhenExpressions) validatePipelineParametersVariables(prefix string, pa
 	for idx, we := range wes {
 		errs = errs.Also(validateStringVariable(we.Input, prefix, paramNames, arrayParamNames).ViaField("input").ViaFieldIndex("when", idx))
 		for _, val := range we.Values {
-			errs = errs.Also(validateStringVariable(val, prefix, paramNames, arrayParamNames).ViaField("values").ViaFieldIndex("when", idx))
+			// one of the values could be a reference to an array param, such as, $(params.foo[*])
+			// extract the variable name from the pattern $(params.foo[*]), if the variable name matches with one of the array params
+			// validate the param as an array variable otherwise, validate it as a string variable
+			if arrayParamNames.Has(strings.TrimSuffix(strings.TrimPrefix(val, "$("+prefix+"."), "[*])")) {
```

Comment:
we're checking if the variable is an array param in two places is slightly different ways, wondering if we can create a common function to check for array param that can be used in both places

https://github.com/tektoncd/pipeline/blob/232ec0e2f9c554f6901951761c73f350207d7e55/pkg/apis/pipeline/v1beta1/when_types.go#L72

https://github.com/tektoncd/pipeline/blob/232ec0e2f9c554f6901951761c73f350207d7e55/pkg/apis/pipeline/v1beta1/when_validation.go#L84

Reply:
Yup, it's very similar but the first expression is returning `params.arrayParam1` from `$(params.arrayParam1[*])` while the second expression is returning `arrayParam1` from `$(params.arrayParam1[*])`.

The `arrayReplacements` in the first expression hold a list of array parameters with `params` prefix while the `arrayReplacements` hold the list of parameters without such prefix.

I have created a common function to return just the parameter name which works for the second expression and added `prefix` back for the first expression.

Validation and applying replacements works with a different patterns because the `apply.go` adds the `prefix` before creating a list of parameter names:

https://github.com/tektoncd/pipeline/blob/2e5780d80a1dd68b0bdd7a494cbb0edf2529a929/pkg/reconciler/pipelinerun/resources/apply.go#L44

We can refactor this further (in a separate PR) to have a consistent validation and replacements routines.

Reply:
thanks for the refactor, looks great!